src/pkg/auth/client.go

```diff
@@ -0,0 +1,249 @@
+	return &client{
+		clientID: clientID,
+		issuer:   issuer,
+	}
+}
+
+func (c client) Authorize(redirectURI string, response ResponseType, opts ...AuthorizeOption) (AuthorizeResult, error) {
+	var as AuthorizeOptions
+	for _, o := range opts {
+		o(&as)
+	}
+
+	result, _ := url.Parse(c.issuer + "/authorize")
+	state := uuid.NewString()
+	values := url.Values{
+		"client_id":     {c.clientID},
+		"state":         {state},
+		"redirect_uri":  {redirectURI},
+		"response_type": {string(response)},
+		// "scope":         {"read:org user:email"}, TODO: add scope AuthorizeOption
+		// "login":         {";TODO: from state file"},
+	}
+	if as.provider != "" {
+		values.Set("provider", as.provider)
+	}
+	var verifier string
+	if as.pkce && response == "code" {
+		pkce, err := GeneratePKCE(64)
+		if err != nil {
+			return AuthorizeResult{}, err
+		}
+		values.Set("code_challenge_method", string(pkce.Method))
+		values.Set("code_challenge", pkce.Challenge)
+		verifier = pkce.Verifier
+	}
+	result.RawQuery = values.Encode()
+	return AuthorizeResult{
+		state:    state,
+		verifier: verifier,
+		url:      *result,
+	}, nil
+}
+
+/**
+ * Exchange the code for access and refresh tokens.
+ */
+func (c client) Exchange(code string, redirectURI string, verifier string) (ExchangeSuccess, error) {
+	body := url.Values{
+		"code":          {code},
+		"redirect_uri":  {redirectURI},
+		"grant_type":    {"authorization_code"},
+		"client_id":     {c.clientID},
+		"code_verifier": {verifier},
+	}
+
+	resp, err := http.PostForm(c.issuer+"/token", body)
+	if err != nil {
+		return ExchangeSuccess{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return ExchangeSuccess{}, errors.New("invalid authorization code: " + resp.Status)
+	}
+
+	var tokens Tokens
+	if err := json.NewDecoder(resp.Body).Decode(&tokens); err != nil {
+		return ExchangeSuccess{}, err
+	}
+
+	return ExchangeSuccess{
+		Tokens: tokens,
+	}, nil
+}
+
+/**
+ * Refreshes the tokens if they have expired.
+ */
+func (c client) Refresh(refresh string, opts ...RefreshOption) (RefreshSuccess, error) {
+	var rs RefreshOptions
+	for _, o := range opts {
+		o(&rs)
+	}
+	if rs.access != "" {
+		var claims jwt.RegisteredClaims
+		_, _, err := new(jwt.Parser).ParseUnverified(rs.access, &claims)
+		if err != nil {
+			return RefreshSuccess{}, errors.New("invalid access token")
+		}
+		// allow 30s window for expiration
+		if claims.ExpiresAt.Unix() > time.Now().Unix()+30 {
+			return RefreshSuccess{
+				Tokens: &Tokens{
+					AccessToken:  rs.access,
+					RefreshToken: refresh,
+				},
+			}, nil
+		}
+	}
+
+	body := url.Values{
+		"grant_type":    {"refresh_token"},
+		"refresh_token": {refresh},
+	}
+	resp, err := http.PostForm(c.issuer+"/token", body)
+	if err != nil {
+		return RefreshSuccess{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return RefreshSuccess{}, errors.New("invalid refresh token")
+	}
```

Comment:
This feels like a greedy error message.

The server probably returns a very specific status code if the refresh token is truly invalid

Could we be more specific by handling non-OK status codes with something like this?

```suggestion
	if resp.StatusCode == http.StatusBadRequest {
		// not sure if HTTP 400 is the expected code, but seems reasonable
		return RefreshSuccess{}, errors.New("invalid refresh token")
	} else if resp.StatusCode != http.StatusOK {
		// a more generic error message here
		return RefreshSuccess{}, errors.new("failed to refresh access token" + resp.Status)
	}
```

Reply:
Now parsing error JSON

Reply:
This has not been resolved. I will still see an error message that says "invalid refresh token:" if the server returns HTTP 5xx.